04-Games/src/js/games/FallingStones.js

```diff
@@ -1,5 +1,193 @@
+                NAME:"easy", 
+                parameters: {
+                    "stone_speed" : 3,
+                    "bullet_time_diff" : 300,
+                    "diff_intensifies" : 0.05
+                },
+            },{
+                NAME: "medium", 
+                parameters: {
+                    "stone_speed" : 5,
+                    "bullet_time_diff" : 400,
+                    "diff_intensifies" : 0.1
+                },
+            },{
+                NAME: "hard", 
+                parameters: {
+                    "stone_speed" : 7,
+                    "bullet_time_diff" : 400,
+                    "diff_intensifies" : 0.2
+                },
+            },
+        ];
+    }
+
+    bindControls()
+    {
+        this.inputBinding = {
+            "left" : this.left.bind(this),
+            "right" : this.right.bind(this),
+            "up" : this.create_bullet.bind(this)
+        };
+    }
+
+    left(bool)
+    {
+        this.player.vx = -bool * 5;
+    }
+
+    right(bool)
+    {
+        this.player.vx = bool * 5;
+    }
+
+    collision_detection()
+    {
+        for(let stone of this.stones)
+        {
+            if(this.bullets.length === 0)
+            {
+                break;
+            }
+
+            for(let bullet of this.bullets)
+            {
+                //wenn das Projektil unterhalb des Steins ist
+                //sind auch alle folgenden Projektile drunter
+                if(bullet.y > stone.y + 100)
+                {
+                    break;
+                }
+
+                if(GameObject.rectangleCollision(stone, bullet))
+                {
+                    this.bullets.splice(this.bullets.indexOf(bullet), 1);
+                    this.stones.splice(this.stones.indexOf(stone), 1);
+                    this.points++;
+                    break;
+                }
+            }
+        }
+    }
+
+    create_bullet(bool)
+    {
+        if(!bool || Date.now() - this.bullet_timer < this.bullet_time_diff) return;
+
+        let x = this.player.x + this.player.width/2 - 5;
+        let y = this.player.y - 10;
+        this.bullets.push(new MovableGameObject(x, y, 10, 10, "#6bd26b", 0, -10));
+        this.bullet_timer = Date.now();
+    }
+
+    update_bullets(ctx)
+    {
+        for(let bullet of this.bullets)
+        {
+            bullet.update(ctx);
+        }
+
+        //es muss nur das oberste Projektil überprüft werden
+        if(this.bullets.length > 0 && this.bullets[0].y + 10 < 0)
+        {
+            this.bullets.shift();
+        }
+    }
+
+    update_stones(ctx)
+    {
+        for(let stone of this.stones)
+        {
+            stone.update(ctx);
+        }
+
+        if(this.stones.length > 0 && this.stones[0].y - 100 > ctx.canvas.height)
+        {
+            this.stones.shift();
+            
+            if(!(--this.life))
+            {
+                this.gameOver = true;
+
+                if(this.gameOverText.length < 3)
```

Comment:
Das kannst du dir sparen, in Javascript kannst du auch einfach immer gleich `this.gameOverText[2] = "..."` schreiben. 
Wenn [2] nicht existiert wird es halt erstellt wenn du was reinschreibst.